midware/middleware.go

```diff
@@ -166,13 +175,32 @@ type traceRequestHeader struct {
 
 func (h *traceRequestHeader) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	var reqid string
+	precedenceHeaders := []string{DefaultTraceHeader, DefaultAzureHeader, DefaultAWSHeader}
+
 	if h.allow {
 		reqid = r.Header.Get(h.header)
+		for _, header := range precedenceHeaders {
+			headerValue := r.Header.Get(header)
+
+			if headerValue != "" {
+				h.header = header
+				reqid = headerValue
+				break
+			}
+		}
 	}
 	if reqid == "" {
 		reqid = uuid.New().String()
 		r.Header.Set(h.header, reqid)
 	}
+
+	// Always set DefaultTraceHeader on request and response since a lot of
+	// logging is hard-coded to use this header.
+	if h.header != DefaultTraceHeader {
+		r.Header.Set(DefaultTraceHeader, reqid)
+		w.Header().Set(DefaultTraceHeader, reqid)
```

Comment:
in the case where `h.header == DefaultTraceHeader` will this still set the response header?

Reply:
it was already setting it before